pandas/io/formats/style.py

```diff
@@ -1589,6 +1688,170 @@ def pipe(self, func: Callable, *args, **kwargs):
+
+    def _translate_tooltips(self, styler_data, uuid, d):
+        """
+        Mutate the render dictionary to allow for tooltips:
+
+        - Add `<span>` HTML element to each data cells `display_value`. Ignores
+          headers.
+        - Add table level CSS styles to control pseudo classes.
+
+        Parameters
+        ----------
+        styler_data : DataFrame
+            Underlying ``Styler`` DataFrame used for reindexing.
+        uuid : str
+            The underlying ``Styler`` uuid for CSS id.
+        d : dict
+            The dictionary prior to final render
+        """
+        self.tt_data = (
+            self.tt_data.reindex_like(styler_data)
+            .dropna(how="all", axis=0)
+            .dropna(how="all", axis=1)
+        )
+        if self.tt_data.empty:
+            return d
+
+        name = self.class_name
+
+        def _pseudo_css(row: int, col: int, text: str):
+            """
+            For every table data-cell that has a valid tooltip (not None, NaN or
+            empty string) must create two pseudo CSS entries for the specific
+            <td> element id which are added to overall table styles:
+            an on hover visibility change and a content change
+            dependent upon the user's chosen display string.
+
+            For example:
+                [{"selector": "T__row1_col1:hover .pd-t",
+                 "props": [("visibility", "visible")]},
+                {"selector": "T__row1_col1 .pd-t::after",
+                 "props": [("content", "Some Valid Text String")]}]
+
+            Parameters
+            ----------
+            row : int
+                The row index of the specified tooltip string data
+            col : int
+                The col index of the specified tooltip string data
+            text : str
+                The textual content of the tooltip to be displayed in HTML.
+
+            Returns
+            -------
+            pseudo_css : List
+            """
+            return [
+                {
+                    "selector": "#T_"
+                    + uuid
+                    + "row"
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f":hover .{name}",
+                    "props": [("visibility", "visible")],
+                },
+                {
+                    "selector": "#T_"
+                    + uuid
+                    + "row"
+                    + str(row)
+                    + "_col"
+                    + str(col)
+                    + f" .{name}::after",
+                    "props": [("content", f'"{text}"')],
+                },
+            ]
+
+        mask = (self.tt_data.isna()) | (self.tt_data.eq(""))  # empty string = no ttip
+        self.table_styles = [
+            style
+            for sublist in [
+                _pseudo_css(i, j, str(self.tt_data.iloc[i, j]))
+                for i, rn in enumerate(self.tt_data.index)
```

Comment:
Are rn / cn on the next line even used? If not should skip enumerate and just iterate over the length